hw2/memoize.py

```diff
@@ -0,0 +1,103 @@
+from collections import OrderedDict
+from dataclasses import dataclass
+from functools import update_wrapper
+from typing import Hashable, Generic, TypeVar, Protocol, Callable, OrderedDict as OrderedDictType, Tuple, Dict
+
+K = TypeVar("K", bound=Hashable)
+
+V = TypeVar("V")
+
+
+class Cache(Generic[K, V]):
+    def __init__(self, size: int):
+        """
+        Cache is a KV-storage, that stores only a fixed number of KV-pairs. Implements FIFO.
+        :param size: the number of KV-pairs to store
+        """
+        self.size = size
+        self.data: OrderedDictType[K, V] = OrderedDict()
+
+    def __setitem__(self, key: K, value: V) -> None:
+        """
+        Put a KV-pair into storage
+        """
+        self.data[key] = value
+        if len(self.data) >= self.size:
+            self.data.popitem(last=False)
+
+    def __getitem__(self, key: K) -> V:
+        """
+        Get a value from storage
+        """
+        return self.data[key]
+
+    def __contains__(self, key: K) -> bool:
+        """
+        Check whether the key is in the storage
+        """
+        return key in self.data
+
+
+def freeze_dict(dictionary: Dict[K, V]) -> Tuple[Tuple[K, V], ...]:
+    """
+    Converts a dict into tuple of KV-pairs
+    """
+    return tuple((k, v) for k, v in dictionary.items())
```

Comment:
Should we sort `items` since it is a dictionary? I could be wrong, but we can get a different order for items depending on the different things.